### DISCUSSION TITLE
Docker Guide

### DISCUSSION THREAD

**Discussion by oc013** [hlky/stable-diffusion#93] [General]
Overview
-------
The current implementation is intended to create the bare minimum environment to run the code in the repository. Minimal information is baked into the image to allow for easy updates without rebuilding.

The Dockerfile will create an image on your system that has Nvidia drivers, CUDA and Miniconda installed.   The `entrypoint.sh` script handles the remaining requirements of the application such as installing the conda env, downloading required model files and launching the webgui when the container is started.

2 volumes are created to cache the conda and pip dependencies and reduce needing to recreate/download them on container restarts/rebuilds.

The code from the repository is mounted in the container at `/sd`.  Image output will be found in `./outputs` on your system as expected.

Requirements
-------
* Nvidia GPU that supports CUDA 11.3+
* Docker and docker compose v2+
* [nvidia-container-toolkit](https://docs.nvidia.com/datacenter/cloud-native/container-toolkit/install-guide.html#docker)

Setup
-------
```
git clone https://github.com/hlky/stable-diffusion.git
cd stable-diffusion
docker compose up
```
Initial installation will take some time, probably 10-20 mins depending on your system and internet connection.

You can stop the container with `CTRL-C` and restart it later by simply running `docker compose up`

Windows Setup (WIP)
-------
* Follow the guide to install [Docker Desktop WSL 2 backend](https://docs.docker.com/desktop/windows/wsl/)
* Verify gpu support by opening WSL and running `nvidia-smi` in the command line

**Make sure that your git is not converting newlines into Windows format differently from the repository**
```
git config --global core.autocrlf input
```

Please comment if you got everything running successfully and any special steps needed. 

Customization
-------
There is now an `.env_docker` file that contains some parameters you can use to customize your setup.  Generally suggest using the defaults for initial setup then you can play with the variables to your liking.  For lower ram systems you might need to use one of the optimized modes to complete your setup e.g. `WEBUI_ARGS=--optimized`

```
# Force miniconda to attempt to update on every container restart
# instead only when changes are detected
CONDA_FORCE_UPDATE=false

# Validate the model files on every container restart
# (useful to set to false after you're sure the model files are already in place)
VALIDATE_MODELS=true

#Automatically relaunch the webui on crashes
WEBUI_RELAUNCH=false

#Pass cli arguments to webui.py e.g:
#WEBUI_ARGS=--gpu=1 --esrgan-gpu=1 --gfpgan-gpu=1
WEBUI_ARGS=
```

Updating
-------
```
docker compose down
git pull
docker compose up --build
```
This will only rebuild layers that have changed in the Dockerfile. Other environment changes should be handled by `entrypoint.sh` and then any new code for the webgui will be launched.

Troubleshooting
-------
A helper script has been provided to clean the docker environment of the container, image and volumes. Please inspect `docker-reset.sh` for the relevant commands in bash if you need to recreate the steps on another OS.

**Comment by MrSaad** [comment]
great work! 
just curious, is the latest compiled version hosted anywhere? would save alot of time if we can just `docker pull`, instead of building from scratch

**Comment by mohammedalsayegh** [comment]
It would be very helpful to have the ability to delete, back up, and restore volumes and images. I had networking issues. So, I wrote few scripts to store volumes and image to other machine. Someone might find these useful if they need to make a quick restoring point or multi testing deployment environment on another machines.

(busybox or ubuntu image are required for volume backup and resorting)

**For volumes (backup and resorting):**
This to backup volumes into two tgz files, ``backup_vol.sh``
```
echo \ && read -p "Press key to continue.. " -n1 -s
echo "backup stable diffusion conda env"
time docker run --rm -v stable-diffusion_conda_env:/volume busybox sh -c 'tar -cOzf - volume' > stable-diffusion_conda_env.tgz

echo \ && read -p "Press key to continue.. " -n1 -s
echo "backup diffusion conda env installation"
time docker run --rm -v stable-diffusion_root_profile:/volume busybox sh -c 'tar -cOzf - volume' > stable-diffusion_root_profile.tgz
```

This to restore the two backup volumes , ``restore_vol.sh``

```
echo \ && read -p "Press key to continue.. " -n1 -s
echo "stable diffusion conda env installation"
docker run -v stable-diffusion_conda_env:/dbdata --name dbstore2 ubuntu /bin/bash
time docker run --rm --volumes-from dbstore2 -v $(pwd):/backup ubuntu bash -c "cd /dbdata && tar xf /backup/stable-diffusion_conda_env.tgz --strip 1 -C /dbdata"
docker stop dbstore2 && docker rm dbstore2

echo \ && read -p "Press key to continue.. " -n1 -s
echo "stable diffusion root profile installation"
docker run -v stable-diffusion_root_profile:/dbdata --name dbstore2 ubuntu /bin/bash
time docker run --rm --volumes-from dbstore2 -v $(pwd):/backup ubuntu bash -c "cd /dbdata && tar xf /backup/stable-diffusion_root_profile.tgz --strip 1 -C /dbdata"
docker stop dbstore2 && docker rm dbstore2

echo done!
echo \ && read -p "Press key to continue.. " -n1 -s
```
 
**For image (save and load):**
For saving image,
``` docker save -o stable-diffusion_stable-diffusion.tar stable-diffusion_stable-diffusion```
For importing image,
```docker load -i stable-diffusion.tar```

**Comment by Synnoid** [comment]
Running this on a headless linux server (with gpus)
I get: 

`sd  | entrypoint.sh: Launching...
sd  | Traceback (most recent call last):
sd  |   File "scripts/webui.py", line 3, in <module>
sd  |     from frontend.frontend import draw_gradio_ui
sd  |   File "/sd/frontend/frontend.py", line 3, in <module>
sd  |     from tkinter.filedialog import askopenfilename
sd  |   File "/opt/conda/envs/ldm/lib/python3.8/tkinter/__init__.py", line 36, in <module>
sd  |     import _tkinter # If this fails your Python may not be configured for Tk
sd  | ImportError: libX11.so.6: cannot open shared object file: No such file or directory`

Any idea?

**Reply by keslerm** [reply]
Add libx11-6 to the Dockerfile apt-get install line and rebuild the container

**Reply by oc013** [reply]
Sad that that was even merged into the codebase 
https://github.com/hlky/stable-diffusion-webui/issues/562
